### ISSUE TITLE
2.x: #delay() operator emits items non-deterministically

### ISSUE THREAD

**Issue by autonomousapps** [ReactiveX/RxJava#5446]
I am using RxJava 2.1.0.

**tl;dr: I think this is a real issue because sometimes the test case is single-threaded (as expected), but sometimes it is multi-threaded. This non-determinism seems problematic.**

I had a flaky test in my test suite. I have boiled it down to the following reproducible (failing) test case:

```java
    @Test
    public void test() throws Exception {
        // Looping to demonstrate flakiness
        for (int i = 0; i < 10; i++) {
            Single<String> source = Single.create(new ErrorsProvider());

            source.observeOn(Schedulers.trampoline()) // Schedulers.io() in production
                    .retryWhen(delayedRetry())
                    .test()
                    .assertResult("Success");
        }
    }

    private static class ErrorsProvider implements SingleOnSubscribe<String> {

        private AtomicInteger count = new AtomicInteger(0);

        @Override public void subscribe(@NonNull SingleEmitter<String> e) throws Exception {
            if (count.getAndIncrement() == 0) {
                e.onError(new Throwable("fail"));
            } else {
                e.onSuccess("Success");
            }
        }
    }

    private static Function<Flowable<Throwable>, Publisher<?>> delayedRetry() {
        return errors -> errors.flatMap(error -> {
            if (error.getMessage().equals("fail")) {
                // This delay would be non-zero in production but set to 0 during test runs
                return Flowable.just(0).delay(0, TimeUnit.SECONDS);
            }
            return Flowable.error(error);
        });
    }
```

I would expect this test to pass 100% of the time, but instead it fails every time I run it inside this loop. More to the point, even with 1 iteration, it'll fail on CI regularly.

I have modified my test like so to workaround the issue:

```java
            Single<String> source = Single.create(new ErrorsProvider());

            source.observeOn(Schedulers.trampoline())
                    .retryWhen(delayedRetry())
                    .test()
                    .awaitCount(1) // new line
                    .assertResult("Success");
```

But I feel like this should not be necessary.

In my production code, I have extensive logging enabled, and by reading them carefully I noticed that _sometimes_, after the `retryWhen()` operator was called, subsequent emissions occurred on a new thread. Most of the time, all emissions are called on a single thread (which is what I'd expect when using `Schedulers.trampoline()`). It was at this point I noticed that the `delay()` operator operates by default on `Schedulers.computation()`. I thought perhaps I was simply mis-using the library, and so I changed my delay call to `delay(...).observeOn(Schedulers.trampoline())`, but that had no effect.

**Comment by akarnokd** [comment]
`delay` runs on the computation scheduler and observing it on the trampoline scheduler has no effect. The trampoline scheduler isn't there to pin execution to a particular thread, its purpose is to make sure tasks submitted from multiple threads execute one after the other. You could use the `delay(long, TimeUnit, Scheduler)` overload to override the scheduler and not introduce asynchrony in your tests.

Alternatively, you could use `awaitDone(long, TimeUnit)` to wait for the async flow and then assert on the contents. `awaitCount` waits only for the `onNext` event and your workaround may still fail to receive `onComplete` in time.

**Comment by autonomousapps** [comment]
Thank you for correcting my misunderstanding of some of these issues. (I actually wish the javadoc were as clear as your sentence "The trampoline scheduler isn't there to pin execution to a particular thread, its purpose is to make sure tasks submitted from multiple threads execute one after the other".)

Nevertheless, the non-deterministic behavior seems wrong. Why would the test sometimes pass (when not in a loop), but other times fail? Why would I sometimes observe all code paths on a single thread, and sometimes on two threads?

**Comment by akarnokd** [comment]
The moment you have an async scheduler in there somewhere (via opeator overload default), you open yourself to concurrent behavior thus events may not arrive "immediately" like in a fully synchronous case.

> non-deterministic behavior seems wrong

Concurrent behavior is often non-intuitive and non-deterministic.

> Why would the test sometimes pass (when not in a loop), but other times fail?

Because the test thread and the consuming thread run concurrently and the test thread might get delayed just enough so the other thread has time to signal all of its events. Other times, the other thread may get things done a bit later and the test thread has already at or behind its checks.

> Why would I sometimes observe all code paths on a single thread, and sometimes on two threads?

Concurrent events are serialized into a single thread and if more events are involved, the same thread may deliver all events if the timing is right.

**Comment by autonomousapps** [comment]
Got it! Thank you.
